File: src/test/java/it/marketplace/microservices/common/dto/OrderDtoTest.java
package it.marketplace.microservices.common.dto;

import it.marketplace.microservices.common.enums.RoleEnum;
import it.marketplace.microservices.common.enums.StatusOrderEnum;
import it.marketplace.microservices.common.enums.StatusUserEnum;
import org.junit.Test;

import java.math.BigDecimal;
import java.time.LocalDateTime;


import static org.junit.jupiter.api.Assertions.assertEquals;

public class OrderDtoTest {

    @Test
    public void shouldCreateOrderDtoWithAllFields() {
        // Arrange
        LocalDateTime now = LocalDateTime.now();
        UserDto user = new UserDto(1L, "John", "Doe", "john.doe@email.com", "Main Street", "London", StatusUserEnum.ACTIVE, RoleEnum.CLIENT, now);
        ProductOrderDto productOrder = new ProductOrderDto(1L, "ORD123", "PROD1", BigDecimal.ONE, 10.0, 10.0, now, now);
        // Act
        OrderDto order = new OrderDto(1L, "ORD123", user, List.of(productOrder), StatusOrderEnum.CREATED, null, now, now);
        // Assert
        assertEquals("1", order.getId().toString());
        assertEquals("ORD123", order.getOrderCode());
        assertEquals(user, order.getUser());
        assertEquals(1, order.getProductOrder().size());
        assertEquals(StatusOrderEnum.CREATED, order.getStatus());
        assertEquals(now, order.getOrderDate());
        assertEquals(now, order.getTmsUpdate());
    }

    @Test
    public void shouldSetAndGetRejectReason() {
        // Arrange
        OrderDto order = new OrderDto();
        // Act
        order.setRejectReason("Out of stock");
        // Assert
        assertEquals("Out of stock", order.getRejectReason());
    }
}

File: src/test/java/it/marketplace/microservices/common/dto/OrderDtoTest.kt
package it.marketplace.microservices.common.dto

import it.marketplace.microservices.common.enums.RoleEnum
import it.marketplace.microservices.common.enums.StatusOrderEnum
import it.marketplace.microservices.common.enums.StatusUserEnum
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.time.LocalDateTime
import kotlin.test.assertEquals

class OrderDtoTest {

    @Test
    fun shouldCreateOrderDtoWithAllFields() {
        // Arrange
        val now = LocalDateTime.now()
        val user = UserDto(
            1L,
            "John",
            "Doe",
            "john.doe@email.com",
            "Main Street",
            "London",
            StatusUserEnum.ACTIVE,
            RoleEnum.CLIENT,
            now,
            now
        )
        val productOrder = ProductOrderDto(1L, "ORD123", "PROD1", BigDecimal.ONE, 10.0, 10.0, now, now)
        // Act
        val order = OrderDto(1L, "ORD123", user, listOf(productOrder), StatusOrderEnum.CREATED, null, now, now)
        // Assert
        assertEquals("1", order.id.toString())
        assertEquals("ORD123", order.orderCode)
        assertEquals(user, order.user)
        assertEquals(1, order.productOrder.size)
        assertEquals(StatusOrderEnum.CREATED, order.status)
        assertEquals(now, order.orderDate)
        assertEquals(now, order.tmsUpdate)
    }

    @Test
    fun shouldSetAndGetRejectReason() {
        // Arrange
        val product = ProductOrderDto(
            id = 1L,
            orderCode = "ORD123",
            productCode = "PROD1",
            unitPrice = 10.0,
            quantity = BigDecimal.ONE,
            tmsUpdate = LocalDateTime.now(),
            total = 10.0,
            creationDate = LocalDateTime.now()
        )

        val user = UserDto(
            id = 1L,
            name = "John",
            lastname = "Doe",
            email = "",
            residenceAddress = "Main Street",
            residenceCity = "London",
            status = StatusUserEnum.ACTIVE,
            role = RoleEnum.CLIENT,
            tmsSubscriptionDate = LocalDateTime.now(),
            tmsUpdate = LocalDateTime.now()
        )

        val order = OrderDto(
            id = 1L,
            orderCode = "ORD123",
            user = user,
            productOrder = listOf(product),
            status = StatusOrderEnum.CREATED,
            rejectReason = null,
            orderDate = LocalDateTime.now(),
            tmsUpdate = LocalDateTime.now()
        )
        // Act
        order.rejectReason = "Out of stock"
        // Assert
        assertEquals("Out of stock", order.rejectReason)
    }
}

